In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from os import listdir

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


Loading the dataset


In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Data_Science/emails.csv')

In [ ]:
df.describe()

spam
count  5728.000000
mean      0.238827
std       0.426404
min       0.000000
25%       0.000000
50%       0.000000
75%       0.000000
max       1.000000

In [ ]:
df.head()

text  spam
0  Subject: naturally irresistible your corporate...     1
1  Subject: the stock trading gunslinger  fanny i...     1
2  Subject: unbelievable new homes made easy  im ...     1
3  Subject: 4 color printing special  request add...     1
4  Subject: do not have money , get software cds ...     1

In [ ]:
df.columns

Index(['text', 'spam'], dtype='object')

In [ ]:
df.tail()

text  spam
5723  Subject: re : research and development charges...     0
5724  Subject: re : receipts from visit  jim ,  than...     0
5725  Subject: re : enron case study update  wow ! a...     0
5726  Subject: re : interest  david ,  please , call...     0
5727  Subject: news : aurora 5 . 2 update  aurora ve...     0

In [ ]:
df.info() #check nan values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5728 entries, 0 to 5727
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   text    5728 non-null   object
 1   spam    5728 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 89.6+ KB


In [ ]:
#1.1 -> How many emails are in the dataset?
len(df)

5728

In [ ]:
#1.2 -> How many of the emails are spam? 
len(df[df['spam'] == 1])

1368

In [ ]:
#1.3 -> Which word appears at the beginning of every email in the dataset? 
#Respond as a lower-case word with punctuation removed.
for i in range(len(df)):
  print(df['text'][i].split(' ')[0][:-1].lower())


Streaming output truncated to the last 5000 lines.
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subject
subje

In [ ]:
#1.4 -> Could a spam classifier potentially benefit from including
# the frequency of the word that appears in every email?
#ANS -> Thwe word subject appears in almost every email, hence the answer is yes

In [ ]:
#1.5 -> How many characters are in the longest email in the dataset 
# where longest is measured in terms of the maximum number of characters)?

max = float('-inf')
for i in range(len(df)):
  if len(df['text'][i]) > max:
    max = len(df['text'][i])
print("The maximum number of characters are", max)

The maximum number of characters are 43952


Creating the corpus

In [ ]:
import nltk
nltk.download('stopwords')
nltk.download('punkt')

from nltk.corpus import stopwords
stop_words = set(stopwords.words('english'))
stop_words_maam = ["i", "me", "my", "myself", "we", "our", "ours", "ourselves", "you", "your", "yours", "yourself", "yourselves", "he", "him", "his", "himself", "she", "her", "hers", "herself", "it", "its", "itself", "they", "them", "their", "theirs", "themselves", "what", "which", "who", "whom", "this", "that", "these", "those", "am", "is", "are", "was", "were", "be", "been", "being", "have", "has", "had", "having", "do", "does", "did", "doing", "would", "should", "could", "ought", "i'm", "you're", "he's", "she's", "it's", "we're", "they're", "i've", "you've", "we've", "they've", "i'd", "you'd", "he'd", "she'd", "we'd", "they'd", "i'll", "you'll", "he'll", "she'll", "we'll", "they'll", "isn't", "aren't", "wasn't", "weren't", "hasn't", "haven't", "hadn't", "doesn't", "don't", "didn't", "won't", "wouldn't", "shan't", "shouldn't", "can't", "cannot", "couldn't", "mustn't", "let's", "that's", "who's", "what's", "here's", "there's", "when's", "where's", "why's", "how's", "a", "an", "the", "and", "but", "if", "or", "because", "as", "until", "while", "of", "at", "by", "for", "with", "about", "against", "between", "into", "through", "during", "before", "after", "above", "below", "to", "from", "up", "down", "in", "out", "on", "off", "over", "under", "again", "further", "then", "once", "here", "there", "when", "where", "why", "how", "all", "any", "both", "each", "few", "more", "most", "other", "some", "such", "no", "nor", "not", "only", "own", "same", "so", "than", "too", "very"]
print(len(stopwords.words('english')))

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
179


In [ ]:
print(stop_words)

{'ourselves', 'in', 'when', 'mustn', 'both', 'do', 'so', 'after', 'only', 'weren', 'below', 'each', 'mightn', 'their', 'wouldn', 'and', 'them', 'off', "mightn't", 'll', "haven't", 'ma', "isn't", 'shouldn', 'isn', "won't", 'been', 'having', 'by', 'until', 'once', 'a', 'our', 'itself', "shouldn't", 'the', "mustn't", 'we', 'her', 'm', 'while', 'all', 'his', 'which', 'other', 'there', 'he', 'again', 'such', 'it', "you've", 'who', 'not', 's', "wasn't", 'i', 'through', 'o', "weren't", 'you', 'where', 'nor', 'doesn', 'too', 'same', 'won', 'didn', 'don', 'between', 'have', 'its', 'haven', 'd', 'at', 'hers', 'above', "don't", 'those', 'about', 'hasn', 'further', 'doing', 'wasn', 'because', 'they', 've', 'for', 'than', 'ain', 'how', 'yourself', 'as', "aren't", 'yourselves', 'yours', 'me', 'themselves', 'on', 'just', 'be', "hasn't", "needn't", 'did', 'herself', 'out', "should've", "couldn't", "hadn't", 'an', 'does', 'she', 'being', "shan't", 'this', 'whom', 'against', 'myself', 'or', "you're", 'o

In [ ]:
#creating a new corpus variable called corpus
corpus = df.copy()

In [ ]:
#convert the text to lowercase. 
for i in range(len(corpus)):
  corpus['text'][i] = corpus['text'][i].lower()
corpus.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


text  spam
0  subject: naturally irresistible your corporate...     1
1  subject: the stock trading gunslinger  fanny i...     1
2  subject: unbelievable new homes made easy  im ...     1
3  subject: 4 color printing special  request add...     1
4  subject: do not have money , get software cds ...     1

In [ ]:
import re
for i in range(len(corpus)):
  corpus["text"][i] = re.sub(r'[^\w\s]', '', corpus["text"][i])

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
from nltk.tokenize import word_tokenize

for i in range(len(corpus)):
  corpus['text'][i] = word_tokenize(corpus['text'][i])
  corpus["text"][i] = [w for w in corpus["text"][i] if not w.lower() in stop_words_maam]

corpus.head()  

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


text  spam
0  [subject, naturally, irresistible, corporate, ...     1
1  [subject, stock, trading, gunslinger, fanny, m...     1
2  [subject, unbelievable, new, homes, made, easy...     1
3  [subject, 4, color, printing, special, request...     1
4  [subject, money, get, software, cds, software,...     1

In [ ]:
#Stemming

from nltk.stem import PorterStemmer
from nltk.stem.snowball import SnowballStemmer

ps = PorterStemmer()
sb = SnowballStemmer("english")

for i in range(len(df)):
    stemmed = []
    for ele in corpus['text'][i]:
        ele = ps.stem(ele)
        stemmed.append(ele)
    corpus['text'][i] = stemmed

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [ ]:
corpus['text'][7]

['subject',
 'save',
 'money',
 'buy',
 'get',
 'thing',
 'tri',
 'ciall',
 'yet',
 'can',
 'even',
 'imagin',
 'like',
 'real',
 'man',
 'bed',
 'thing',
 'great',
 'errrectlon',
 'provid',
 'exactiy',
 'want',
 'ciali',
 'lot',
 'advantag',
 'viagra',
 'effect',
 'iast',
 '36',
 'hour',
 'readi',
 'start',
 'within',
 'just',
 '10',
 'minut',
 'can',
 'mix',
 'aicohol',
 'ship',
 'countri',
 'get',
 'right',
 'now']

In [ ]:
for i in range(len(corpus)):
    corpus['text'][i] = ' '.join(corpus['text'][i])

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [ ]:
! pip install textmining3

     |████████████████████████████████| 1.9 MB 14.3 MB/s 


In [ ]:
import textmining

# Initialize for creating term-document matrix
tdm = textmining.TermDocumentMatrix()

for i in range(len(corpus)):
    tdm.add_doc(corpus['text'][i])


In [ ]:
tdm_new = tdm.to_df(cutoff=0)

In [ ]:
tdm_new.shape

(5728, 25677)

In [ ]:
#including terms which appear in 5% 
threshold = 286
sptdm = tdm.to_df(cutoff=threshold)
sptdm.shape

(5728, 348)

In [ ]:
emailsSparse = pd.DataFrame(sptdm)
emailsSparse.head()

subject  corpor  realli  compani  market  full  inform  good  websit  will  \
0        1       1       1        3       4     1       1     2       2     6   
1        1       0       0        0       0     0       0     0       0     0   
2        1       0       0        0       0     0       0     0       1     0   
3        1       0       0        0       0     0       2     0       0     0   
4        1       0       0        0       0     0       0     0       0     0   

   ...  ect  hou  vinc  shirley  enron  kaminski  crenshaw  vkamin  stinson  \
0  ...    0    0     0        0      0         0         0       0        0   
1  ...    0    0     0        0      0         0         0       0        0   
2  ...    0    0     0        0      0         0         0       0        0   
3  ...    0    0     0        0      0         0         0       0        0   
4  ...    0    0     0        0      0         0         0       0        0   

   gibner  
0       0  
1       0  
2       0  
3       0  
4       0  

[5 rows x 348 columns]

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import roc_auc_score

X_train, X_test, y_train, y_test = train_test_split(sptdm ,corpus['spam'], test_size=0.3, random_state = 123)

In [ ]:
cart_classifier = DecisionTreeClassifier()
cart_classifier.fit(X_train, y_train)
print("Training accuracy is ", cart_classifier.score(X_train, y_train))

Training accuracy is  0.9992516837116487


In [ ]:
y_pred = cart_classifier.predict(X_test)
print("Testing accuracy without threshold is ", accuracy_score(y_test, y_pred))

Testing accuracy without threshold is  0.9348458406050029


In [ ]:
Threshold_value = 0.6
y_pred = np.where(cart_classifier.predict_proba(X_test)[:,1] >= Threshold_value, 1, 0)
print("Testing accuracy with threshold is ", accuracy_score(y_test, y_pred))

Testing accuracy with threshold is  0.9348458406050029


In [ ]:
Random_forest = RandomForestClassifier()
Random_forest.fit(X_train, y_train)
print("Training accuracy is ", cart_classifier.score(X_train, y_train))

Training accuracy is  0.9992516837116487


In [ ]:
y_pred = Random_forest.predict(X_test)
print("Testing accuracy without threshold is ", accuracy_score(y_test, y_pred))

Testing accuracy without threshold is  0.9778941244909831


In [ ]:
Threshold_value = 0.5
y_pred = np.where(Random_forest.predict_proba(X_test)[:,1] >= Threshold_value, 1, 0)
print("Testing accuracy with threshold is ", accuracy_score(y_test, y_pred))

Testing accuracy with threshold is  0.9784758580570099


###Conclusion:

  The email dataset had uneven class distribution with 1368 spam emails and 4360 ham emails. In this experiment, I performed text analysis for separating spam email from ham. A spam classifier can potentially benefit from including the frequency of the word that appears in every email. Random Forest model outperformed CART model in terms of Accuracy and AUC. This is because Random forests consist of multiple single trees each based on a random sample of the training data. They are typically more accurate than single decision trees.